pkg/overlay/server.go

```diff
@@ -69,118 +66,48 @@ func (server *Server) BulkLookup(ctx context.Context, reqs *pb.LookupRequests) (
+		UptimeCount:       server.nodeSelectionConfig.UptimeCount,
+		UptimeRatio:       server.nodeSelectionConfig.UptimeRatio,
 	}
 
-	excluded := opts.ExcludedNodes
-	restrictions := opts.GetRestrictions()
-	reputation := server.nodeStats
-
-	var startID storj.NodeID
-	result := []*pb.Node{}
-	for {
-		var nodes []*pb.Node
-		nodes, startID, err = server.populate(ctx, req.Start, maxNodes, restrictions, reputation, excluded)
-		if err != nil {
-			return nil, Error.Wrap(err)
-		}
-
-		resultNodes := []*pb.Node{}
-		usedAddrs := make(map[string]bool)
-		for _, n := range nodes {
-			addr := n.Address.GetAddress()
-			excluded = append(excluded, n.Id) // exclude all nodes on next iteration
-			if !usedAddrs[addr] {
-				resultNodes = append(resultNodes, n)
-				usedAddrs[addr] = true
-			}
-		}
-		if len(resultNodes) <= 0 {
-			break
-		}
-
-		result = append(result, resultNodes...)
-
-		if len(result) >= int(maxNodes) || startID == (storj.NodeID{}) {
-			break
-		}
-
-	}
-
-	if len(result) < int(maxNodes) {
-		return nil, status.Errorf(codes.ResourceExhausted, fmt.Sprintf("requested %d nodes, only %d nodes matched the criteria requested", maxNodes, len(result)))
-	}
-
-	if len(result) > int(maxNodes) {
-		result = result[:maxNodes]
+	filterNodesReq := &FilterNodesRequest{
+		MinReputation:         minStats,
+		MinNodes:              req.GetMinNodes(),
+		Opts:                  req.GetOpts(),
+		NewNodePercentage:     server.nodeSelectionConfig.NewNodePercentage,
+		NewNodeAuditThreshold: server.nodeSelectionConfig.NewNodeAuditThreshold,
 	}
 
-	return &pb.FindStorageNodesResponse{
-		Nodes: result,
-	}, nil
-}
-
-// TODO: nicer method arguments
-func (server *Server) populate(ctx context.Context,
-	startID storj.NodeID, maxNodes int64,
-	minRestrictions *pb.NodeRestrictions,
-	minReputation *pb.NodeStats,
-	excluded storj.NodeIDList) ([]*pb.Node, storj.NodeID, error) {
-
-	// TODO: move the query into db
-	limit := int(maxNodes * 2)
-	nodes, err := server.cache.db.List(ctx, startID, limit)
+	foundNodes, err := server.cache.db.FilterNodes(ctx, filterNodesReq)
 	if err != nil {
-		server.log.Error("Error listing nodes", zap.Error(err))
-		return nil, storj.NodeID{}, Error.Wrap(err)
-	}
-
-	var nextStart storj.NodeID
-	result := []*pb.Node{}
-	for _, v := range nodes {
-		if v == nil {
-			continue
-		}
-
-		nextStart = v.Id
-		if v.Type != pb.NodeType_STORAGE {
-			continue
-		}
-
-		restrictions := v.GetRestrictions()
-		reputation := v.GetReputation()
-
-		if restrictions.GetFreeBandwidth() < minRestrictions.GetFreeBandwidth() ||
-			restrictions.GetFreeDisk() < minRestrictions.GetFreeDisk() ||
-			reputation.GetUptimeRatio() < minReputation.GetUptimeRatio() ||
-			reputation.GetUptimeCount() < minReputation.GetUptimeCount() ||
-			reputation.GetAuditSuccessRatio() < minReputation.GetAuditSuccessRatio() ||
-			reputation.GetAuditCount() < minReputation.GetAuditCount() ||
-			contains(excluded, v.Id) {
-			continue
+		_, ok := status.FromError(err)
```

Comment:
ok isn't saying the status is okay. ok is saying that it was an error from this package. you need to actually check the real return value

Reply:
er, maybe i don't understand what you're trying to do

Reply:
I was also thinking of wrapping the resource exhausted error in a zeebo error, but wasn't sure if that was the right way to go about it since `Error` is already the default error class for overlay errors. so I wasn't sure if I should make like a child error... 😬

Reply:
oh actually I'll just do this, I think this works:
```
	foundNodes, err := server.cache.db.FilterNodes(ctx, filterNodesReq)
	if err != nil {
		stat, ok := status.FromError(err)
		if ok && stat.Code() == codes.ResourceExhausted {
			return &pb.FindStorageNodesResponse{
				Nodes: foundNodes,
			}, err
		}
		return nil, Error.Wrap(err)
	}
```